src/reconcile/rec_write.c

```diff
@@ -2350,19 +2395,104 @@ __rec_split_grow(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t add_len)
-	size_t corrected_page_size, len;
+	size_t corrected_page_size, inuse, len;
 
 	btree = S2BT(session);
 	bm = btree->bm;
 
 	len = WT_PTRDIFF(r->first_free, r->disk_image.mem);
-	corrected_page_size = len + add_len;
+	inuse = (len - r->bnd[r->bnd_next].offset) +
+	    WT_PAGE_HEADER_BYTE_SIZE(btree);
+	corrected_page_size = inuse + add_len;
+
 	WT_RET(bm->write_size(bm, session, &corrected_page_size));
-	WT_RET(__wt_buf_grow(session, &r->disk_image, corrected_page_size));
+	/* Need to account for buffer carrying two chunks worth of data */
+	WT_RET(__wt_buf_grow(session, &r->disk_image, 2 * corrected_page_size));
+
 	r->first_free = (uint8_t *)r->disk_image.mem + len;
-	WT_ASSERT(session, corrected_page_size >= len);
-	r->space_avail = corrected_page_size - len;
+	WT_ASSERT(session, corrected_page_size >= inuse);
+	r->space_avail = corrected_page_size - inuse;
 	WT_ASSERT(session, r->space_avail >= add_len);
+
+	return (0);
+}
+
+/*
+ * __rec_split_write_prev_and_shift_cur --
+ *	Write the previous split chunk to the disk as a page. Shift the contents
+ *	of the current chunk to the start of the buffer, making space for a new
+ *	chunk to be written.
+ *	If the caller asks for a chunk resizing, the boundary between the two
+ *	chunks is readjusted to the minimum split size boundary details stored
+ *	in the previous chunk, letting the current chunk grow at the cost of the
+ *	previous chunk.
+ */
+static int
+__rec_split_write_prev_and_shift_cur(
+    WT_SESSION_IMPL *session, WT_RECONCILE *r, bool resize_chunks)
+{
+	WT_BOUNDARY *bnd_cur, *bnd_prev;
+	WT_BTREE *btree;
+	WT_PAGE_HEADER *dsk, *dsk_tmp;
+	size_t cur_len, prev_len_aligned;
+	uint8_t *dsk_start;
+
+	WT_ASSERT(session, r->bnd_next != 0);
+
+	btree = S2BT(session);
+	bnd_cur = &r->bnd[r->bnd_next];
+	bnd_prev = bnd_cur - 1;
+	dsk = r->disk_image.mem;
+	cur_len = WT_PTRDIFF(r->first_free, dsk) - bnd_cur->offset;
+
+	/*
+	 * Resize chunks if the current is smaller than the minimum, and there
+	 * are details on the minimum split size boundary available in the
+	 * previous boundary details.
+	 */
+	if (resize_chunks &&
+	    cur_len < r->min_split_size && bnd_prev->min_bnd_offset != 0) {
+		bnd_cur->offset = bnd_prev->min_bnd_offset;
+		bnd_cur->entries +=
+		    bnd_prev->entries - bnd_prev->min_bnd_entries;
+		bnd_prev->entries = bnd_prev->min_bnd_entries;
+		bnd_cur->recno = bnd_prev->min_bnd_recno;
+
+		WT_RET(__wt_buf_set(session, &bnd_cur->key,
+		    bnd_prev->min_bnd_key.data, bnd_prev->min_bnd_key.size));
+
+		/* Update current chunk's length */
+		cur_len = WT_PTRDIFF(r->first_free, dsk) - bnd_cur->offset;
+	}
+
+	/*
+	 * Create an interim buffer if not already done to prepare the previous
+	 * chunk's disk image.
+	 */
+	prev_len_aligned = WT_ALIGN(bnd_cur->offset, btree->allocsize);
+	if (r->interim_buf == NULL)
+		WT_RET(__wt_scr_alloc(
+		    session, prev_len_aligned, &r->interim_buf));
+	else
+		WT_RET(__wt_buf_init(
+		    session, r->interim_buf, prev_len_aligned));
```

Comment:
I just realized this is wrong. Reconciliation has to ask the block manager how much space it needs to write a chunk of data, so it  should be something like this:
```
diff --git a/src/reconcile/rec_write.c b/src/reconcile/rec_write.c
index 0ac5b34db..7c0633d2c 100644
--- a/src/reconcile/rec_write.c
+++ b/src/reconcile/rec_write.c
@@ -2431,15 +2431,17 @@ static int
 __rec_split_write_prev_and_shift_cur(
     WT_SESSION_IMPL *session, WT_RECONCILE *r, bool resize_chunks)
 {
+       WT_BM *bm;
        WT_BOUNDARY *bnd_cur, *bnd_prev;
        WT_BTREE *btree;
        WT_PAGE_HEADER *dsk, *dsk_tmp;
-       size_t cur_len, prev_len_aligned;
+       size_t cur_len, len;
        uint8_t *dsk_start;
 
        WT_ASSERT(session, r->bnd_next != 0);
 
        btree = S2BT(session);
+       bm = btree->bm;
        bnd_cur = &r->bnd[r->bnd_next];
        bnd_prev = bnd_cur - 1;
        dsk = r->disk_image.mem;
@@ -2469,13 +2471,12 @@ __rec_split_write_prev_and_shift_cur(
         * Create an interim buffer if not already done to prepare the previous
         * chunk's disk image.
         */
-       prev_len_aligned = WT_ALIGN(bnd_cur->offset, btree->allocsize);
+       len = bnd->cur_offset;
+       WT_RET(bm->write_size(bm, session, &len));
        if (r->interim_buf == NULL)
-               WT_RET(__wt_scr_alloc(
-                   session, prev_len_aligned, &r->interim_buf));
+               WT_RET(__wt_scr_alloc(session, len, &r->interim_buf));
        else
-               WT_RET(__wt_buf_init(
-                   session, r->interim_buf, prev_len_aligned));
+               WT_RET(__wt_buf_init(session, r->interim_buf, len));
 
        dsk_tmp = r->interim_buf->mem;
        memcpy(dsk_tmp, dsk, bnd_cur->offset);
```